### PR TITLE
LibGUI+WindowServer+Taskbar: Add super key event & show start menu when it is pressed

### LINKED ISSUE #6153: TaskBar: Logo key does not open Serenity menu
I think this used to work, but seems like it broke when it was moved to the Taskbar.

### PR DESCRIPTION
**Pull request by dreamhopping**
This PR re-implements the start menu being shown when the super key is pressed via the new ``WM_SuperKeyPressed`` event.
The existing functionality powered by the super key (app switching, tiling, etc.) is still fully functional!

Fixes #6153

### CODE REVIEW

File: Userland/Services/Taskbar/TaskbarWindow.cpp [hunk 1]
```diff
@@ -84,14 +84,15 @@ TaskbarWindow::TaskbarWindow(NonnullRefPtr<GUI::Menu> start_menu)
...
+    m_start_button->set_font(Gfx::FontDatabase::default_bold_font());
+    m_start_button->set_icon_spacing(0);
+    m_start_button->set_fixed_size(80, 22);
     auto app_icon = GUI::Icon::default_icon("ladybug");
-    start_button.set_icon(app_icon.bitmap_for_size(16));
-    start_button.set_menu(m_start_menu);
+    m_start_button->set_icon(app_icon.bitmap_for_size(16));
+    m_start_button->set_menu(m_start_menu);
 
+    main_widget.add_child(*m_start_button);
```
Comment (sin-ack) [inline Userland/Services/Taskbar/TaskbarWindow.cpp:95]: Is there a particular reason you changed the `main_widget.add` part to adding the child later? Just curious.

File: Userland/Services/WindowServer/WindowManager.h [hunk 2]
```diff
@@ -334,6 +335,7 @@ class WindowManager : public Core::Object {
     DoubleClickInfo m_double_click_info;
     int m_double_click_speed { 0 };
     int m_max_distance_for_double_click { 4 };
+    bool m_previous_event_was_super_keydown;
```
Comment (sin-ack) [inline Userland/Services/WindowServer/WindowManager.h:338]:
```suggestion
    bool m_previous_event_was_super_keydown { false };
```
Makes it a little clearer, I think. Not sure if leaving it off causes the variable to be uninitialized, if so that's another plus.

File: Userland/Services/WindowServer/WindowManagerServer.ipc [hunk 3]
```diff
@@ -9,4 +9,6 @@ endpoint WindowManagerServer = 1871
     PopupWindowMenu(i32 client_id, i32 window_id, Gfx::IntPoint screen_position) =|
     SetWindowTaskbarRect(i32 client_id, i32 window_id, Gfx::IntRect rect) =|
     SetAppletAreaPosition(Gfx::IntPoint position) => ()
+
+    SuperPressed() => ()
```
Comment (sin-ack) [inline Userland/Services/WindowServer/WindowManagerServer.ipc:13]:
```suggestion
```
This isn't called anywhere, so it can be removed.

**Review by sin-ack — CHANGES_REQUESTED** [review]
Looks pretty good overall! Just a couple minor things that caught my eye.

File: Userland/Services/Taskbar/TaskbarWindow.cpp (hunk 1 above)
Comment (dreamhopping) [reply Userland/Services/Taskbar/TaskbarWindow.cpp:95]: No, it was just put there out of "style", the creation of the button happens first, then it's added. I can change it if you like :)

**Comment by dreamhopping** [comment]
> Looks pretty good overall! Just a couple minor things that caught my eye.

Thank you, I've committed the two of those and I'll squash them in a bit

**Comment by dreamhopping** [comment]
I've squashed those changes into the original commit and fixed the build error :smile: 

_So many force pushes :sweat_smile:_

**Comment by dreamhopping** [comment]
This PR is ready to be merged, once @sin-ack is happy with the 1st question :)

**Review by awesomekling — CHANGES_REQUESTED** [review]
Please wrap your commit messages at 72 characters per the [contribution guidelines](https://github.com/SerenityOS/serenity/blob/969c73660b81d3bfd80e2b7fab64652888961c1e/CONTRIBUTING.md)

**Comment by dreamhopping** [comment]
Done! Sorry I missed that one, won't happen again :p

**Comment by dreamhopping** [comment]
> This PR is ready to be merged, once @sin-ack is happy with the 1st question :)

Spoke to them in Discord, it's ok 😄 image

**Comment by dreamhopping** [comment]
> Please wrap your commit messages at 72 characters per the [contribution guidelines](https://github.com/SerenityOS/serenity/blob/969c73660b81d3bfd80e2b7fab64652888961c1e/CONTRIBUTING.md)

This has been fixed in the force push a few hours ago :p

File: Userland/Services/WindowServer/WMClientConnection.cpp [hunk 4]
```diff
@@ -24,6 +24,7 @@
  * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
 
+#include <LibGUI/Event.h>
```
Comment (awesomekling) [inline Userland/Services/WindowServer/WMClientConnection.cpp:27]: Is this change actually necessary?

File: Userland/Services/WindowServer/WMClientConnection.h [hunk 5]
```diff
@@ -47,6 +47,7 @@ class WMClientConnection final
     virtual void handle(const Messages::WindowManagerServer::StartWindowResize&) override;
     virtual void handle(const Messages::WindowManagerServer::PopupWindowMenu&) override;
     virtual void handle(const Messages::WindowManagerServer::SetWindowTaskbarRect&) override;
+
```
Comment (awesomekling) [inline Userland/Services/WindowServer/WMClientConnection.h:50]: Please avoid making whitespace-only changes where there's nothing else being changed in the same file.

File: Userland/Services/WindowServer/WMClientConnection.cpp (hunk 4 above)
Comment (dreamhopping) [reply Userland/Services/WindowServer/WMClientConnection.cpp:27]: It isn't, that was an artifact of my old testing... sorry about that, I'll remove it now

File: Userland/Libraries/LibGUI/WindowManagerServerConnection.h [hunk 6]
```diff
@@ -52,6 +52,7 @@ class WindowManagerServerConnection
     virtual void handle(const Messages::WindowManagerClient::WindowIconBitmapChanged&) override;
     virtual void handle(const Messages::WindowManagerClient::WindowRectChanged&) override;
     virtual void handle(const Messages::WindowManagerClient::AppletAreaSizeChanged&) override;
+    virtual void handle(const Messages::WindowManagerClient::SuperPressed&) override;
```
[the remark this answers is not in the document]
Comment (dreamhopping) [reply Userland/Libraries/LibGUI/WindowManagerServerConnection.h:55]: Alright, changing now :p

File: Userland/Services/WindowServer/WMClientConnection.cpp [hunk 7]
```diff
@@ -170,5 +170,4 @@ void WMClientConnection::handle(const Messages::WindowManagerServer::SetWindowTa
     auto& window = *(*it).value;
     window.set_taskbar_rect(message.rect());
 }
-
```
Comment (linusg) [inline Userland/Services/WindowServer/WMClientConnection.cpp:173]: Please don't make whitespace changes in files that are otherwise not changed, especially not like this where there actually should be an empty line :)

File: Userland/Services/WindowServer/WindowManager.cpp [hunk 8]
```diff
@@ -377,6 +378,17 @@ void WindowManager::tell_wms_applet_area_size_changed(const Gfx::IntSize& size)
     });
 }
 
+void WindowManager::tell_wms_super_key_pressed()
+{
+    for_each_window_manager([&](WMClientConnection& conn) {
```
Comment (linusg) [inline Userland/Services/WindowServer/WindowManager.cpp:383]: do we actually need to capture anything? this only uses `conn`, which is passed as an argument.

Comment (dreamhopping) [reply Userland/Services/WindowServer/WindowManager.cpp:383]: Oops... I'll fix that now 😅

File: Userland/Services/WindowServer/WMClientConnection.cpp [hunk 9]
```diff
@@ -171,4 +171,4 @@ void WMClientConnection::handle(const Messages::WindowManagerServer::SetWindowTa
     window.set_taskbar_rect(message.rect());
 }
 
-}
+}
```
Comment (linusg) [inline Userland/Services/WindowServer/WMClientConnection.cpp:174]: This is now removing a newline at the end of the file (which will fail the build soon..)

**Review by linusg — CHANGES_REQUESTED** [review]

File: Userland/Services/WindowServer/WMClientConnection.cpp (hunk 9 above)
Comment (dreamhopping) [reply Userland/Services/WindowServer/WMClientConnection.cpp]: Yeah... :sweat_smile: just fixed that, sorry

File: Userland/Services/WindowServer/WindowManager.cpp [hunk 10]
```diff
@@ -1211,7 +1223,17 @@ void WindowManager::event(Core::Event& event)
             return;
         }
 
-        if (MenuManager::the().current_menu()) {
+        if (key_event.type() == Event::KeyDown && key_event.key() == Key_Super) {
+            m_previous_event_was_super_keydown = true;
```
Comment (awesomekling) [inline Userland/Services/WindowServer/WindowManager.cpp:1227]: After `m_previous_event_was_super_keydown` is set to `true`, it stays `true` even if the next event is e.g a `MouseDown` event. That makes the value a lie, and means that any number of unrelated mouse events can happen in between super keydown/keyup.

If you look at 7ee6c66ee95ad151c24d34c259464db03fb7ea6b where this functionality was originally implemented, notice that `m_previous_event_is_key_down_logo` was cleared upon any mouse event other than `MouseMove`.

**Review by awesomekling — CHANGES_REQUESTED** [review]

File: Userland/Services/WindowServer/WindowManager.cpp (hunk 10 above)
Comment (dreamhopping) [reply Userland/Services/WindowServer/WindowManager.cpp:1227]: Oh, okay.. I've added that and I'll just test it before pushing. Thank you :^)
